<a href="https://colab.research.google.com/github/wbwatkinson/udemy_numpy_stack/blob/main/numpy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Section 1 of:

https://www.udemy.com/deep-learning-prerequisites-the-numpy-stack-in-python/

https://deeplearningcourses.com/c/deep-learning-prerequisites-the-numpy-stack-in-python

## Arrays vs Lists

In [1]:
import numpy as np

In [2]:
L = [1,2,3]

In [3]:
A = np.array([1,2,3])

In [4]:
for e in L:
  print(e)

1
2
3


In [5]:
for e in A:
  print(e)

1
2
3


In [6]:
L.append(4)

In [7]:
L

[1, 2, 3, 4]

In [8]:
# A.append(4) # doesn't work

In [9]:
L + [5]

[1, 2, 3, 4, 5]

In [10]:
A + np.array([4])

array([5, 6, 7])

In [11]:
A + np.array([4,5,6])

array([5, 7, 9])

In [12]:
# A + np.array([4,5]) # doesn't work

In [13]:
2 * A

array([2, 4, 6])

In [14]:
2 * L

[1, 2, 3, 4, 1, 2, 3, 4]

In [15]:
L + L

[1, 2, 3, 4, 1, 2, 3, 4]

In [16]:
L2 = []
for e in L:
  L2.append(e + 3)

In [17]:
L2

[4, 5, 6, 7]

In [18]:
L2 = [e + 3 for e in L]

In [19]:
L2

[4, 5, 6, 7]

In [20]:
L2 = []
for e in L:
  L2.append(e**2)

In [21]:
L2

[1, 4, 9, 16]

In [22]:
A**2

array([1, 4, 9])

In [23]:
np.sqrt(A)

array([1.        , 1.41421356, 1.73205081])

In [24]:
np.log(A)

array([0.        , 0.69314718, 1.09861229])

In [25]:
np.exp(A)

array([ 2.71828183,  7.3890561 , 20.08553692])

In [26]:
np.tanh(A)

array([0.76159416, 0.96402758, 0.99505475])

## The Dot Product

$$ a \cdot b = a^T b = \sum_{d=1}^D a_d b_d $$

In [27]:
a = np.array([1,2])
b = np.array([3,4])

In [28]:
dot = 0
for e, f in zip(a, b):
  dot += e*f
dot

11

In [29]:
# use integer index
dot = 0
for i in range(len(a)):
  dot += a[i] * b[i]
dot

11

In [30]:
a * b

array([3, 8])

In [31]:
np.sum(a * b)

11

In [32]:
(a * b).sum()

11

In [33]:
np.dot(a, b)

11

In [34]:
a.dot(b)

11

In [35]:
b.dot(a)

11

In [36]:
a @ b

11

$$ a^T b = \left\lVert a \right\rVert \left\lVert b \right\rVert \cos \theta_{ab}$$

$$ \cos \theta_{ab} = \frac{a^T b}{\left\lVert a \right\rVert \left\lVert b \right\rVert} $$

$$ \left\lVert a \right\rVert = \sqrt{\sum_{d=1}^D a_d^2 } $$

In [37]:
amag = np.sqrt((a * a).sum())
amag

2.23606797749979

In [38]:
amag = np.linalg.norm(amag)
amag

2.23606797749979

In [39]:
cosangle = a.dot(b) / (np.linalg.norm(a) * np.linalg.norm(b))

In [40]:
cosangle

0.9838699100999074

In [41]:
angle = np.arccos(cosangle)
angle

0.17985349979247847

## Speed Test

In [42]:
## speed comparison ##
import numpy as np
from datetime import datetime

# note: you can also use %timeit

a = np.random.randn(100)
b = np.random.randn(100)
T = 100000

def slow_dot_product(a, b):
  result = 0
  for e, f in zip(a, b):
    result += e*f
  return result

t0 = datetime.now()
for t in range(T):
  slow_dot_product(a, b)
dt1 = datetime.now() - t0

t0 = datetime.now()
for t in range(T):
  a.dot(b)
dt2 = datetime.now() - t0

print("dt1 / dt2:", dt1.total_seconds() / dt2.total_seconds())

dt1 / dt2: 71.08106445793013


## Matrices

In [43]:
L = [[1,2], [3,4]]
L

[[1, 2], [3, 4]]

In [44]:
L[0]

[1, 2]

In [45]:
L[0][1]

2

In [46]:
A = np.array([[1,2],[3,4]])
A

array([[1, 2],
       [3, 4]])

In [47]:
A[0][1]

2

In [48]:
A[0,1]

2

In [49]:
A[:,0] # returns the column at index 0 (select every row of column 0)

array([1, 3])

In [50]:
A.T

array([[1, 3],
       [2, 4]])

In [51]:
np.exp(A)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [52]:
np.exp(L) # pass a list into numpy operation

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [53]:
B = np.array([[1,2,3],[4,5,6]])
B

array([[1, 2, 3],
       [4, 5, 6]])

In [54]:
A.dot(B)

array([[ 9, 12, 15],
       [19, 26, 33]])

In [55]:
# A.dot(B.T) # doesn't work, not aligned

In [56]:
np.linalg.det(A)

-2.0000000000000004

In [57]:
np.linalg.inv(A)

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [58]:
np.linalg.inv(A).dot(A)

array([[1.00000000e+00, 0.00000000e+00],
       [1.11022302e-16, 1.00000000e+00]])

In [59]:
np.trace(A)

5

In [60]:
np.diag(A) # get the diagonal elements of A

array([1, 4])

In [61]:
np.diag([1,4]) # create a diagonal matrix with 1 and 4 on the diagonal

array([[1, 0],
       [0, 4]])

In [62]:
np.linalg.eig(A) # two arrays: eigen values, eigen vectors in a matrix

(array([-0.37228132,  5.37228132]), array([[-0.82456484, -0.41597356],
        [ 0.56576746, -0.90937671]]))

In [63]:
Lam, V = np.linalg.eig(A)

In [64]:
V[:,0] * Lam[0] == A @ V[:,0] # eigen vector times eigen value should equal matrix times eigen vector

array([ True, False])

In [65]:
V[:,0] * Lam[0], A @ V[:,0]

(array([ 0.30697009, -0.21062466]), array([ 0.30697009, -0.21062466]))

In [66]:
np.allclose(V[:,0] * Lam[0], A @ V[:,0])

True

In [67]:
np.allclose(V @ np.diag(Lam), A @ V)

True

# Solving Linear Systems

The admission fee at a small fair is \\$1.50 for children and \\$4.00 for adults. On a certain day, 2200 people enter the fair, and \\$5050 is collected. How many children and how many adults attended?

\begin{align}
  x_1 + x_2 & = 2200 \\
  1.5x_1 + 4x_2 & = 5500
\end{align}

$$ x = \left ( \begin{matrix} x_1 \\ x_2 \end{matrix} \right ) , 
A = \left ( \begin{matrix} 1 & 1 \\ 1.5 & 4 \end{matrix} \right ) ,
b = \left ( \begin{matrix} 2200 \\ 5050 \end{matrix} \right ) $$

$$ Ax = b \Leftrightarrow x = A^{-1}b $$

In [68]:
A = np.array([[1,1], [1.5, 4]])
b = np.array([2200, 5050])

In [69]:
np.linalg.solve(A, b) # faster
np.linalg.inv(A).dot(b) # don't use this, it's slower

array([1500.,  700.])

# Generating Data

In [70]:
np.zeros((2,3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [71]:
np.ones((2,3))

array([[1., 1., 1.],
       [1., 1., 1.]])

In [72]:
10 * np.ones((2,3))

array([[10., 10., 10.],
       [10., 10., 10.]])

In [73]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [74]:
np.random.random()

0.9739361675364091

In [75]:
np.random.random((2, 3)) # uniform [0,1] distribution

array([[0.07480892, 0.4535755 , 0.27616302],
       [0.17907705, 0.94127473, 0.4577829 ]])

In [76]:
np.random.randn(2,3) # gaussian (normal) distribution mean = 0, variance = 1

array([[ 1.26628165e+00, -4.90935555e-04,  5.16494651e-01],
       [-6.28922619e-01,  6.20152832e-01,  7.83885130e-01]])

In [77]:
R = np.random.randn(10000)

In [78]:
R.mean()

0.007728137417581059

In [79]:
np.mean(R)

0.007728137417581059

In [80]:
R.var()

0.9968685959051978

In [81]:
R.std() # standard deviation is the square root of the variance

0.9984330703182852

In [82]:
R = np.random.randn(10000,3)

In [83]:
R.mean(axis = 0) # the mean of each column

array([-0.00830576, -0.01493858, -0.01011861])

In [84]:
R.mean(axis = 1) # the mean of each row

array([ 0.3214464 , -0.39900122, -0.49932406, ..., -0.49924568,
       -0.02784231,  0.77375251])

In [85]:
R.mean(axis=1).shape

(10000,)

In [86]:
np.cov(R) # treats each column as a vector observation

array([[ 0.97167476,  1.17837247, -0.73171622, ..., -1.14010586,
         0.68729556, -0.66929676],
       [ 1.17837247,  2.18169171, -0.87807591, ..., -1.54115743,
         1.21712499, -1.63838159],
       [-0.73171622, -0.87807591,  0.55113104, ...,  0.85659529,
        -0.5128287 ,  0.49380384],
       ...,
       [-1.14010586, -1.54115743,  0.85659529, ...,  1.37112168,
        -0.88723188,  0.95943617],
       [ 0.68729556,  1.21712499, -0.5128287 , ..., -0.88723188,
         0.68167889, -0.89478717],
       [-0.66929676, -1.63838159,  0.49380384, ...,  0.95943617,
        -0.89478717,  1.36907121]])

In [87]:
np.cov(R).shape # analog of variance with vectors is covariance

(10000, 10000)

In [88]:
np.cov(R.T) # take transpose first to treat each row as an observation

array([[ 1.00400541, -0.01484353,  0.00694562],
       [-0.01484353,  0.98680753,  0.00420037],
       [ 0.00694562,  0.00420037,  0.98560652]])

In [89]:
np.cov(R, rowvar = False) # passing rowvar = False has same effect as transposing R

array([[ 1.00400541, -0.01484353,  0.00694562],
       [-0.01484353,  0.98680753,  0.00420037],
       [ 0.00694562,  0.00420037,  0.98560652]])

In [90]:
np.random.randint(0,10,size=(3,3))

array([[7, 7, 6],
       [1, 8, 0],
       [5, 7, 9]])

In [91]:
np.random.choice(10, size=(3,3))

array([[7, 9, 6],
       [5, 2, 7],
       [5, 4, 1]])

# Numpy Exercise

Do a speed test for matrix multiplication

In [92]:
## speed comparison ##
import numpy as np
from datetime import datetime

# note: you can also use %timeit

a = np.random.randint(0,10, size=(3,3))
b = np.random.randint(0,10, size=(3,3))
T = 100000

def list_comp_matrix_multiplication(a, b):
  return [[sum([a*b for (a,b) in zip(row, col)]) for col in zip(*b)] for row in a]

def for_loop_matrix_multiply(a, b):
  result = np.zeros((a.shape[0], b.shape[1]))

  for i, row in enumerate(a):
    for j, col in enumerate(b.T):
      # result[i,j] = slow_dot_product(row, col) #row @ col
      result[i, j] = sum([row*col for (row, col) in zip(row,col)])

  return result


def slow_dot_product(a, b):
  result = 0
  for e, f in zip(a, b):
    result += e*f
  return result

t0 = datetime.now()
for t in range(T):
  for_loop_matrix_multiply(a,b)
dt1 = datetime.now() - t0

t0 = datetime.now()
for t in range(T):
  a @ b
dt2 = datetime.now() - t0

print("dt1 / dt2:", dt1.total_seconds() / dt2.total_seconds())
a, b, a.dot(b), a @ b, for_loop_matrix_multiply(a, b)

dt1 / dt2: 30.34996871759271


(array([[5, 2, 4],
        [4, 7, 4],
        [8, 0, 6]]), array([[7, 8, 2],
        [3, 8, 5],
        [2, 4, 8]]), array([[ 49,  72,  52],
        [ 57, 104,  75],
        [ 68,  88,  64]]), array([[ 49,  72,  52],
        [ 57, 104,  75],
        [ 68,  88,  64]]), array([[ 49.,  72.,  52.],
        [ 57., 104.,  75.],
        [ 68.,  88.,  64.]]))